openfecli/commands/gather.py

```diff
@@ -296,36 +353,42 @@ def _write_raw(legs:dict, writer:Callable, allow_partial=True):
             for repeat in repeats:
                 for m, u in repeat:
                     if m is None:
-                        m, u = 'NaN', 'NaN'
+                        m, u = FAIL_STR, FAIL_STR
                     else:
                         m, u = format_estimate_uncertainty(m.m, u.m)
                     writer.writerow([simtype, *ligpair, m, u])
 
+def _write_dg_mle(legs: dict, writer: Callable, allow_partial: bool) -> None:
+    """Compute and write out DG values for the given legs.
 
-def _write_dg_raw(legs:dict, writer:Callable,  allow_partial):  # pragma: no-cover
-    writer.writerow(["leg", "ligand_i", "ligand_j", "DG(i->j) (kcal/mol)",
-                     "uncertainty (kcal/mol)"])
-    for ligpair, vals in sorted(legs.items()):
-        for simtype, (m, u) in sorted(vals.items()):
-            if m is None:
-                m, u = 'NaN', 'NaN'
-            else:
-                m, u = format_estimate_uncertainty(m.m, u.m)
-            writer.writerow([simtype, *ligpair, m, u])
-
-
-def _write_dg_mle(legs:dict, writer:Callable, allow_partial:bool):
+    Parameters
+    ----------
+    legs : dict
+        Dict of legs to write out.
+    writer : Callable
+        The CSV writer to use.
+    allow_partial : bool
+        If ``True``, no error will be thrown for incomplete or invalid results,
+        and DGs will be reported for whatever valid results are found.
+    """
     import networkx as nx
     import numpy as np
     from cinnabar.stats import mle
+
     DDGs = _get_ddgs(legs, allow_partial=allow_partial)
     MLEs = []
-    # 4b) perform MLE
+    expected_ligs = []
+
+    # perform MLE
     g = nx.DiGraph()
     nm_to_idx = {}
     DDGbind_count = 0
     for ligA, ligB, DDGbind, bind_unc, DDGhyd, hyd_unc in DDGs:
-        if DDGbind is None:
+        for lig in (ligA, ligB):
+            if lig not in expected_ligs:
+                expected_ligs.append(lig)
+
+        if DDGbind is None or DDGbind == FAIL_STR:
```

Comment:
If you have one or more failed edges, there is a risk that you might end up with a disconnected graph. We should have a is_connected check on `g` below before we call the mle. If it's disconnected then for now I would suggest we just fail saying it is disconnected.

Reply:
good call! I added basic handling here, but it's something I'd like to handle more nicely when we do a rewrite.